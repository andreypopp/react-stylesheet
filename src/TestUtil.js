/**
 * @flow
 */

import {CSSClassJoin, CSSClass} from './StyleRepresentation';

export const snapshotSerializer = {
  // $FlowFixMe: ...
  print(val: mixed, serialize, indent) {
    if (val instanceof CSSClassJoin) {
      const styles = val.styles.map(style => indent(serialize(style)));
      return ['CSSClassJoin [', ...styles, ']'].join('\n');
    } else if (val instanceof CSSClass) {
      if (val.repr != null) {
        const repr = val.repr.map(rule =>
          [
            `CSSClass ${rule.selector} {`,
            ...rule.props.map(prop => indent(prop)),
            `}`,
          ].join('\n'),
        );
        return repr.join('\n');
      } else {
        return val.className;
      }
    }
  },

  test(val: mixed) {
    return val instanceof CSSClassJoin || (val instanceof CSSClass && val.repr != null);
  },
};
