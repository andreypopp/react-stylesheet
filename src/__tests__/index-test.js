/**
 * @flow
 */

import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as ReactStylesheet from '../index';
import * as TestUtil from '../TestUtil';

// $FlowFixMe: ...
expect.addSnapshotSerializer(TestUtil.snapshotSerializer);

const stylesheet = ReactStylesheet.createStylesheet({
  base: {
    color: 'red',
  },
  some: {
    color: 'black',
  },
});

const stylesheetWithRTL = ReactStylesheet.createStylesheet({
  base: {
    color: 'red',
  },
  rightToLeft: {
    color: 'black',
  },
});

test('toClassName()', function() {
  expect(ReactStylesheet.toClassName(stylesheet, {})).toMatchSnapshot();
  expect(ReactStylesheet.toClassName(stylesheet, {base: true})).toMatchSnapshot();
  expect(ReactStylesheet.toClassName(stylesheet, {base: false})).toMatchSnapshot();
  expect(ReactStylesheet.toClassName(stylesheet, {some: true})).toMatchSnapshot();
  expect(ReactStylesheet.toClassName(stylesheet, {some: false})).toMatchSnapshot();
  expect(
    ReactStylesheet.toClassName(stylesheet, {}, {rightToLeft: true}),
  ).toMatchSnapshot('Component-<HASH> RTL');
  expect(
    ReactStylesheet.toClassName(stylesheet, {some: true}, {rightToLeft: true}),
  ).toMatchSnapshot();
  expect(ReactStylesheet.toClassName(stylesheetWithRTL, {})).toMatchSnapshot();
  expect(
    ReactStylesheet.toClassName(stylesheetWithRTL, {}, {rightToLeft: true}),
  ).toMatchSnapshot();
  expect(
    ReactStylesheet.toClassName(
      stylesheetWithRTL,
      {rightToLeft: true},
      {rightToLeft: true},
    ),
  ).toMatchSnapshot();
  expect(
    ReactStylesheet.toClassName(stylesheetWithRTL, {rightToLeft: true}),
  ).toMatchSnapshot();
});

beforeEach(function() {
  ReactStylesheet.staticEnvironment.dispose();
});

afterEach(function() {
  ReactStylesheet.staticEnvironment.dispose();
});

test('styleComponent() inserts a stylesheet into DOM', function() {
  const FancyDiv = ReactStylesheet.styleComponent('div', {
    base: {
      color: 'red',
    },
  });
  expect(ReactStylesheet.staticEnvironment.sheet.cssRules.length).toBe(1);
});

describe('<Element />', function() {
  test('<Element color={...} />', function() {
    const component = renderer.create(<ReactStylesheet.Element color="red" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element hover={...} />', function() {
    const component = renderer.create(<ReactStylesheet.Element hover={{color: 'red'}} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element color={...} hover={...} />', function() {
    const component = renderer.create(
      <ReactStylesheet.Element color="black" hover={{color: 'red'}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element focus={...} hover={...} />', function() {
    const component = renderer.create(
      <ReactStylesheet.Element focus={{color: 'black'}} hover={{color: 'red'}} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element className={...} />', function() {
    const component = renderer.create(<ReactStylesheet.Element className="bootstrap" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('<Element color="red" className={...} />', function() {
    const component = renderer.create(
      <ReactStylesheet.Element color="red" className="bootstrap" />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
