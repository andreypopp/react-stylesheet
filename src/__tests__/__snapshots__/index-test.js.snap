// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Element /> <Element className={...} /> 1`] = `
<div
  Component="div"
  className={
    CSSClassJoin [
      "bootstrap"
    ]
  }
/>
`;

exports[`<Element /> <Element color="red" className={...} /> 1`] = `
<div
  Component="div"
  className={
    CSSClassJoin [
      CSSClass base {
        color: red
      }
      "bootstrap"
    ]
  }
/>
`;

exports[`<Element /> <Element color={...} /> 1`] = `
<div
  Component="div"
  className={
    CSSClassJoin [
      CSSClass base {
        color: red
      }
    ]
  }
/>
`;

exports[`<Element /> <Element color={...} hover={...} /> 1`] = `
<div
  Component="div"
  className={
    CSSClassJoin [
      CSSClass base {
        color: black
      }
      CSSClass base:hover {
        color: red
      }
    ]
  }
/>
`;

exports[`<Element /> <Element focus={...} hover={...} /> 1`] = `
<div
  Component="div"
  className={
    CSSClassJoin [
      CSSClass base:focus {
        color: black
      }
      CSSClass base:hover {
        color: red
      }
    ]
  }
/>
`;

exports[`<Element /> <Element hover={...} /> 1`] = `
<div
  Component="div"
  className={
    CSSClassJoin [
      CSSClass base:hover {
        color: red
      }
    ]
  }
/>
`;

exports[`Component-<HASH> RTL 1`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
  "RTL"
]
`;

exports[`toClassName() 1`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
]
`;

exports[`toClassName() 2`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
]
`;

exports[`toClassName() 3`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
]
`;

exports[`toClassName() 4`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
  CSSClass some {
    color: black
  }
]
`;

exports[`toClassName() 5`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
]
`;

exports[`toClassName() 6`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
  "RTL"
  CSSClass some {
    color: black
  }
]
`;

exports[`toClassName() 7`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
]
`;

exports[`toClassName() 8`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
  "RTL"
  CSSClass rightToLeft {
    color: black
  }
]
`;

exports[`toClassName() 9`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
  "RTL"
  CSSClass rightToLeft {
    color: black
  }
]
`;

exports[`toClassName() 10`] = `
CSSClassJoin [
  CSSClass base {
    color: red
  }
  CSSClass rightToLeft {
    color: black
  }
]
`;
